### PR TITLE
Use pinned arrays only in IO and unpinned arrays elsewhere

### PR DESCRIPTION
**Pull request by adithyaov**

### CODE REVIEW

**Comment by harendra-kumar** [comment]
Perf regressions:

```
Benchmark                                                        default(0)(μs) default(1) - default(0)(%)
---------------------------------------------------------------- -------------- --------------------------
All.Data.Array/o-1-space.elimination./=                                  106.35                     +59.00
All.Data.Array/o-1-space.elimination.writeLastN.1                        129.51                     +55.19
All.Data.Array/o-1-space.elimination.foldl'                              187.14                     +33.50
All.Data.Array/o-1-space.transformation.scanl1'                          222.60                     +32.56
All.Data.Array/o-1-space.generation.show                               46788.00                     +17.06
All.Data.Array/o-1-space.transformation.map                              219.19                      +6.74

Data.Array.Mut(cpuTime)
Benchmark                                                    default(0)(μs) default(1) - default(0)(%)
------------------------------------------------------------ -------------- --------------------------
All.Data.Array.Mut/o-1-space.elimination.toStreamRev                 146.57                     +44.21
All.Data.Array.Mut/o-1-space.strip (> 0)                              59.42                     +34.85
All.Data.Array.Mut/o-1-space.partitionBy (> value/2)                  81.63                     +25.88
All.Data.Array.Mut/o-1-space.elimination.readRev                     158.36                     +20.34
All.Data.Array.Mut/o-1-space.generation.writeN . intFromTo            77.46                     +14.48
All.Data.Array.Mut/o-1-space.transformation.scanl1'                  216.57                      +9.03
All.Data.Array.Mut/o-1-space.transformationX4.map                    667.89                      +7.93

Benchmark                                                                       default(0)(μs) default(1) - default(0)(%)
------------------------------------------------------------------------------- -------------- --------------------------
All.Data.Parser/o-1-space.parseMany (take 1)                                             39.35                    +103.58
All.Data.Parser/o-1-space.takeWhileP                                                     79.37                    +100.11
All.Data.Parser/o-1-space/filesystem.S.parseMany (Fold.take 1 Fold.sum)              130180.00                     +81.96
All.Data.Parser/o-1-space.dropWhile                                                      49.77                     +59.06
All.Data.Parser/o-1-space.takeStartBy                                                    49.35                     +58.20
All.Data.Parser/o-1-space.takeEndBy_                                                     50.26                     +57.96
All.Data.Parser/o-1-space.splitApBefore                                                  64.94                     +53.93
All.Data.Parser/o-1-space.alt2                                                          119.59                     +31.91
All.Data.Parser/o-1-space.splitApAfter                                                   99.92                     +19.89
All.Data.Parser/o-1-space.parseIterate (take 1)                                        1283.75                     +14.27
All.Data.Parser/o-1-space.parseMany                                                      70.08                     +13.98
All.Data.Parser/o-1-space.concatSequence                                               3962.68                     +12.98
All.Data.Parser/o-1-space.sepBy1 (words)                                                946.96                     +11.14
All.Data.Parser/o-1-space.many (wordBy even)                                            189.47                     +10.28
All.Data.Parser/o-1-space.manyTill                                                      987.15                      +8.88
All.Data.Parser/o-1-space.parseIterate (take all)                                      1118.80                      +7.16
All.Data.Parser/o-1-space.sepBy (words)                                                 996.88                      +6.99
All.Data.Parser/o-1-space.parseMany (groupBy (==))                                     1635.40                      +6.28
All.Data.Parser/o-n-heap.listEqBy                                                      1185.48                      +5.97

Data.Stream.StreamK(cpuTime)
Benchmark                                                                        default(0)(μs) default(1) - default(0)(%)
-------------------------------------------------------------------------------- -------------- --------------------------
All.Data.Stream.StreamK/o-1-space.transformationX4.mapM                                 4516.07                     +17.99
All.Data.Stream.StreamK/o-1-space.transformationX4.mapMSerial                           4504.50                     +17.10
All.Data.Stream.StreamK/o-1-space.mixedX2.take-drop                                     3554.29                     +15.54
All.Data.Stream.StreamK/o-n-stack.transformation.intersperse                              16.12                     +15.42
All.Data.Stream.StreamK/o-1-space.concat.concatMapBy serial (1 of n)                    5819.55                     +15.29
All.Data.Stream.StreamK/o-1-space.filteringX4.take-all                                  4992.34                     +14.67
All.Data.Stream.StreamK/o-1-space.filteringX4.dropWhile-false                           2331.48                     +14.24
All.Data.Stream.StreamK/o-n-stack.transformationX4.intersperse                             1.06                     +14.23
All.Data.Stream.StreamK/o-n-stack.transformation.interspersePure                          15.01                     +14.10

FileSystem.Handle(cpuTime)
Benchmark                                                                                    default(0)(ms) default(1) - default(0)(%)
-------------------------------------------------------------------------------------------- -------------- --------------------------
All.FileSystem.Handle/o-1-space.reduce/read/chunks.S.groupsOf 1 FL.sum                                67.02                    +127.74
All.FileSystem.Handle/o-1-space.copy.FH.write . FH.read                                              124.20                     +29.92
All.FileSystem.Handle/o-1-space.copy/read.rawToNull                                                  155.89                     +27.48
All.FileSystem.Handle/o-1-space.copy.FH.writeWith . FH.readWith                                      123.70                     +26.96
All.FileSystem.Handle/o-1-space.reduce/read/chunks.S.chunksOf 1000                                   161.68                     +24.03
All.FileSystem.Handle/o-1-space.copy/read.rawToFile                                                  280.73                     +15.03
All.FileSystem.Handle/o-1-space.reduce/read/chunks.S.foldManyPost (FL.take 104857600 FL.sum)         320.56                     +12.20
```

**Comment by harendra-kumar** [comment]
There are two classes of regressions:

1. Regressions because we simply changed the array type from pinned to unpinned. Unpinned memory is copied by GC which may have some impact on benchmarks. We can verify that by using a pinned array creation API explicitly in the becnhmarks.
2. Regressions because we are creating unpinned arrays in the middle of a computation and then these unpinned arrays need to be copied to pinned arrays during IO. Only benchmarks that are doing IO to files/network will be affected by this class of problems.

**Comment by harendra-kumar** [comment]
New results, after using pinned chunksOf:

```
Data.Array(cpuTime)
Benchmark                                                        default(0)(μs) default(1) - default(0)(%)
---------------------------------------------------------------- -------------- --------------------------
All.Data.Array/o-1-space.generation.writeN . IsList.fromList            2927.95                     +16.88
All.Data.Array/o-1-space.transformationX4.map                            688.02                     +10.91
All.Data.Array/o-1-space.transformationX4.scanl'                         670.26                     +10.85
All.Data.Array/o-1-space.transformation.map                              217.33                     +10.78
All.Data.Array/o-1-space.transformationX4.scanl1'                        614.21                     +10.59

Data.Array.Mut(cpuTime)
Benchmark                                                    default(0)(μs) default(1) - default(0)(%)
------------------------------------------------------------ -------------- --------------------------
All.Data.Array.Mut/o-1-space.transformation.scanl'                   239.05                     +22.48
All.Data.Array.Mut/o-1-space.generation.fromList . intFromTo        1067.83                     +10.86
All.Data.Array.Mut/o-1-space.transformationX4.scanl1'                642.14                     +10.10
All.Data.Array.Mut/o-1-space.elimination.toStreamRev                 157.18                      +9.76
All.Data.Array.Mut/o-1-space.partitionBy (> 0)                       104.18                      +9.36
All.Data.Array.Mut/o-1-space.modifyIndices (+ 1)                     158.35                      +8.17
All.Data.Array.Mut/o-1-space.transformationX4.map                    719.51                      +7.40

Data.Parser(cpuTime)
Benchmark                                                                       default(0)(μs) default(1) - default(0)(%)
------------------------------------------------------------------------------- -------------- --------------------------
All.Data.Parser/o-1-space.parseMany groupRollingBy (bound groups)                      1493.69                     +14.52
All.Data.Parser/o-1-space.parseMany/Unfold/1000 arrays/take 1                          1018.11                     +13.49
All.Data.Parser/o-1-space.parseIterate (take all)                                      1123.98                     +12.22
All.Data.Parser/o-n-heap.listEqBy                                                      1086.28                     +11.80
All.Data.Parser/o-n-heap.manyAlt                                                       3497.37                     +11.29
All.Data.Parser/o-1-space.sepByAll (words)                                              295.93                     +10.96
All.Data.Parser/o-1-space.concatSequence                                               3963.15                      +8.94
All.Data.Parser/o-1-space.parseMany groupRollingByEither (Right)                        279.82                      +7.56
All.Data.Parser/o-1-space/filesystem.S.parseMany (Fold.take 1 Fold.sum)              134111.00                      +7.46
All.Data.Parser/o-1-space.many (wordBy even)                                            196.94                      +7.45
All.Data.Parser/o-1-space.sepBy (words)                                                 942.58                      +7.37
All.Data.Parser/o-1-space.parseMany groupRollingBy (1 group)                             49.77                      +7.29
All.Data.Parser/o-n-heap.sepBy1                                                        2103.07                      +6.64
All.Data.Parser/o-n-heap.monad16                                                        323.03                      +6.32
All.Data.Parser/o-1-space.manyTill                                                      881.89                      +6.28
All.Data.Parser/o-n-heap.lookAhead                                                       58.74                      +5.67
All.Data.Parser/o-1-space.parseMany (groupBy (==))                                     1550.73                      +5.41

Data.Parser.ParserK(cpuTime)
Benchmark                                         default(0)(μs) default(1) - default(0)(%)
------------------------------------------------- -------------- --------------------------
All.Data.Parser.ParserK/o-n-heap.sequence_               1890.92                     +19.80
All.Data.Parser.ParserK/o-n-heap.sepBy1                  3099.13                     +11.57
All.Data.Parser.ParserK/o-1-space.alt2                    234.14                      +8.55
All.Data.Parser.ParserK/o-1-space.alt16                   694.35                      +7.71
All.Data.Parser.ParserK/o-1-space.monad2                  195.27                      +7.26
All.Data.Parser.ParserK/o-1-space.monad16                 186.09                      +6.44
All.Data.Parser.ParserK/o-n-heap.manyAlt                 9491.92                      +5.34

FileSystem.Handle(cpuTime)
Benchmark                                                                                    default(0)(ms) default(1) - default(0)(%)
-------------------------------------------------------------------------------------------- -------------- --------------------------
All.FileSystem.Handle/o-1-space.copy.FH.write . FH.read                                              134.07                     +17.95
All.FileSystem.Handle/o-1-space.copy.FH.writeWith . FH.readWith                                      136.07                     +15.84
All.FileSystem.Handle/o-1-space.reduce/toBytes.US.lines . SS.decodeLatin1                             88.42                     +11.82

Unicode.Stream(cpuTime)
Benchmark                                                                                            default(0)(ms) default(1) - default(0)(%)
---------------------------------------------------------------------------------------------------- -------------- --------------------------
All.Unicode.Stream/o-1-space.decode-encode.encodeUtf8 . parseMany parseCharUtf8 (1/10)                       250.32                     +12.58
All.Unicode.Stream/o-1-space.decode-encode.encodeUtf8 . decodeUtf8 (1/10)                                    190.44                      +8.14
```

better than before. Need to check if the variance is just run-to-run variance or actual change.

File: core/src/Streamly/Data/Array.hs [hunk 1]
```diff
@@ -111,12 +111,9 @@ import Prelude hiding (read)
...
 -- == Pinned and Unpinned Arrays
 --
--- The array type can use both pinned and unpinned memory under the hood.
--- Currently the array creation APIs create arrays in pinned memory but it will
--- change to unpinned in future releases. The change should not affect users
--- functionally unless they are directly accessing the internal memory of the
--- array via internal APIs. As of now unpinned arrays can be created using
--- unreleased APIs.
+-- The array type can use both pinned and unpinned memory under the hood.  The
+-- arrays are unpinned by default and are only pinned While doing any I/O
```
Comment (harendra-kumar) [inline core/src/Streamly/Data/Array.hs:115]: While => while

**Review by harendra-kumar — COMMENTED** [review]
looks good. need to double check the benchmarks.

**Comment by harendra-kumar** [comment]
On local machine the cpuTime variation is much less in most benchmarks. Some FileSystem.Handle benchmarks have consistent significant regression:

```
FileSystem.Handle(cpuTime)
Benchmark                                                                                    default(0)(ms) default(1) - default(0)(%)
-------------------------------------------------------------------------------------------- -------------- --------------------------
All.FileSystem.Handle/o-1-space.copy.FH.writeWith . FH.readWith                                      120.11                     +47.76
All.FileSystem.Handle/o-1-space.copy.FH.write . FH.read                                              124.08                     +26.69
All.FileSystem.Handle/o-1-space.reduce/toBytes.US.lines . SS.decodeLatin1                             84.56                     +16.83
All.FileSystem.Handle/o-1-space.reduce/read.US.lines . SS.decodeLatin1                               110.80                     +16.25

Data.ParserK(cpuTime)
Benchmark                                  default(0)(μs) default(1) - default(0)(%)
------------------------------------------ -------------- --------------------------
All.Data.ParserK/o-1-space.alt8                    418.44                     +27.61
All.Data.ParserK/o-n-heap.sequence               13240.10                     +17.47
All.Data.ParserK/o-1-space.alt16                   663.10                     +15.99
```

**Comment by adithyaov** [comment]
Hmmm.... These shouldn't have regressions. Let me look into these.

**Review by harendra-kumar — CHANGES_REQUESTED** [review]

**Pull request closed without merging**